quantized elu: require observation

vkuzo opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #40284 graph mode: add hardswish inplace handling
* #40111 quantized elu: graph mode handling
* #40104 quantized elu: eager mode QAT handling
* #40103 quantized elu: eager mode static handling
* **#40100 quantized elu: require observation**

Summary:

ELU has a range of [-1, inf]. In the original PR which added
the quantized operator we decided to pass the quantization params
from the input.  However, it makes more sense to require observation
for this op.

This PR changes the API to require observation. Next PRs in this stack
will add the eager and graph mode handling.

Test Plan:

```
python test/test_quantization.py TestQuantizedOps.test_qelu
```

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D22075083](https://our.internmc.facebook.com/intern/diff/D22075083)

vkuzo committed:
quantized elu: require observation

Summary:

ELU has a range of [-1, inf]. In the original PR which added
the quantized operator we decided to pass the quantization params
from the input.  However, it makes more sense to require observation
for this op.

This PR changes the API to require observation. Next PRs in this stack
will add the eager and graph mode handling.

Test Plan:

```
python test/test_quantization.py TestQuantizedOps.test_qelu
```

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

vkuzo committed:
Update on "quantized elu: require observation"


Summary:

ELU has a range of [-1, inf]. In the original PR which added
the quantized operator we decided to pass the quantization params
from the input.  However, it makes more sense to require observation
for this op.

This PR changes the API to require observation. Next PRs in this stack
will add the eager and graph mode handling.

Test Plan:

```
python test/test_quantization.py TestQuantizedOps.test_qelu
```

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

vkuzo committed:
Update on "quantized elu: require observation"


Summary:

ELU has a range of [-1, inf]. In the original PR which added
the quantized operator we decided to pass the quantization params
from the input.  However, it makes more sense to require observation
for this op.

This PR changes the API to require observation. Next PRs in this stack
will add the eager and graph mode handling.

Test Plan:

```
python test/test_quantization.py TestQuantizedOps.test_qelu
```

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 commented on test/quantization/test_quantized_op.py:
```diff
@@ -306,35 +306,24 @@ def test_qrelu_leaky(self, X, alpha):
            alpha=st.floats(0.01, 10.0, allow_nan=False, allow_infinity=False))
     def test_qelu(self, X, alpha):
         X, (scale, zero_point, torch_type) = X
+        output_scale = 0.5
+        output_zero_point = 1
```
can we use `_test_activation_function` here?

vkuzo replied:
took a quick look, this would require `_test_activation_function` making that func a bit more flexible, so if possible I'd recommend that to be a separate PR, to keep this one single purposed.

jerryzh168 commented on torch/nn/quantized/functional.py:
```diff
@@ -472,8 +472,8 @@ def threshold(input, threshold, value):
 
-def elu(input, alpha=1., inplace=False, scale=None, zero_point=None):
-    # type: (Tensor, Optional[float], bool, Optional[float], Optional[int]) -> Tensor
+def elu(input, scale, zero_point, alpha=1.):
```
why the functional has less arguments than the quantized op? missing `scale` and `input_scale`

vkuzo replied:
I agree, it's a bit confusing.  The quantized python op is following the floating point python op (https://pytorch.org/docs/stable/_modules/torch/nn/modules/activation.html#ELU).  The 'scale' and 'input_scale' arguments seen in c++ for ELU for both fp and quant are not related to quantization, and also not exposed to Python.  I believe they are there because the kernel is reused with other functions, which do use these params.

jerryzh168 replied:
I see

jerryzh168 commented on torch/nn/quantized/functional.py:
```diff
@@ -482,25 +482,13 @@ def elu(input, alpha=1., inplace=False, scale=None, zero_point=None):
         input: quantized input
+        scale, zero_point: Scale and zero point of the output tensor.
         alpha: the :math:`\alpha` value for the ELU formulation. Default: 1.0
         inplace: Inplace modification of the input tensor
```
comment for inplace can be removed

vkuzo committed:
Update on "quantized elu: require observation"


Summary:

ELU has a range of [-1, inf]. In the original PR which added
the quantized operator we decided to pass the quantization params
from the input.  However, it makes more sense to require observation
for this op.

This PR changes the API to require observation. Next PRs in this stack
will add the eager and graph mode handling.

Test Plan:

```
python test/test_quantization.py TestQuantizedOps.test_qelu
```

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D22075083](https://our.internmc.facebook.com/intern/diff/D22075083)

[ghstack-poisoned]

raghuramank100 commented on test/quantization/test_quantized_op.py:
```diff
@@ -306,35 +306,24 @@ def test_qrelu_leaky(self, X, alpha):
         dqY_hat = dqX.clone()
         dqY_hat[dqX < 0] = alpha * (torch.exp(dqY_hat[dqX < 0]) - 1.)
-        qY_hat = torch.quantize_per_tensor(dqY_hat, scale=scale, zero_point=zero_point,
+        qY_hat = torch.quantize_per_tensor(dqY_hat, scale=output_scale, zero_point=output_zero_point,
```
Why are we not calling F.elu on the floating point tensor here?

vkuzo replied:
Happy to change (ideally in a separate PR) if this is important to fix.

raghuramank100 commented on test/quantization/test_quantize_script.py:
```diff
@@ -2285,7 +2285,6 @@ def __init__(self):
                 self.adaptive_avg_pool1d = torch.nn.AdaptiveAvgPool1d((1))
                 self.adaptive_avg_pool2d = torch.nn.AdaptiveAvgPool2d((1, 1))
                 self.adaptive_avg_pool3d = torch.nn.AdaptiveAvgPool3d((1, 1, 1))
-                self.elu = torch.nn.ELU()
```
I understand why we are removing the module here, but dont see where we test that graph mode quantization works correctly for elu (i.e observers are inserted ).

vkuzo replied:
it's added back in the correct way further up the PR stack.  Since this functionality is not released yet, I split the diffs for easier review.

vkuzo committed:
Update on "quantized elu: require observation"


Summary:

ELU has a range of [-1, inf]. In the original PR which added
the quantized operator we decided to pass the quantization params
from the input.  However, it makes more sense to require observation
for this op.

This PR changes the API to require observation. Next PRs in this stack
will add the eager and graph mode handling.

Test Plan:

```
python test/test_quantization.py TestQuantizedOps.test_qelu
```

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D22075083](https://our.internmc.facebook.com/intern/diff/D22075083)

[ghstack-poisoned]

vkuzo committed:
Update on "quantized elu: require observation"


Summary:

ELU has a range of [-1, inf]. In the original PR which added
the quantized operator we decided to pass the quantization params
from the input.  However, it makes more sense to require observation
for this op.

This PR changes the API to require observation. Next PRs in this stack
will add the eager and graph mode handling.

Test Plan:

```
python test/test_quantization.py TestQuantizedOps.test_qelu
```

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D22075083](https://our.internmc.facebook.com/intern/diff/D22075083)

[ghstack-poisoned]

vkuzo committed:
Update on "quantized elu: require observation"


Summary:

ELU has a range of [-1, inf]. In the original PR which added
the quantized operator we decided to pass the quantization params
from the input.  However, it makes more sense to require observation
for this op.

This PR changes the API to require observation. Next PRs in this stack
will add the eager and graph mode handling.

Test Plan:

```
python test/test_quantization.py TestQuantizedOps.test_qelu
```

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D22075083](https://our.internmc.facebook.com/intern/diff/D22075083)

[ghstack-poisoned]

vkuzo committed:
Update on "quantized elu: require observation"


Summary:

ELU has a range of [-1, inf]. In the original PR which added
the quantized operator we decided to pass the quantization params
from the input.  However, it makes more sense to require observation
for this op.

This PR changes the API to require observation. Next PRs in this stack
will add the eager and graph mode handling.

Test Plan:

```
python test/test_quantization.py TestQuantizedOps.test_qelu
```

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D22075083](https://our.internmc.facebook.com/intern/diff/D22075083)

[ghstack-poisoned]

vkuzo committed:
Update on "quantized elu: require observation"


Summary:

ELU has a range of [-1, inf]. In the original PR which added
the quantized operator we decided to pass the quantization params
from the input.  However, it makes more sense to require observation
for this op.

This PR changes the API to require observation. Next PRs in this stack
will add the eager and graph mode handling.

Test Plan:

```
python test/test_quantization.py TestQuantizedOps.test_qelu
```

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D22075083](https://our.internmc.facebook.com/intern/diff/D22075083)

[ghstack-poisoned]
